### ISSUE TITLE
[ERROR] Unable to migrate certificate storage

### ISSUE THREAD

**Issue by OleVik** [mholt/caddy#902]
1. What version of Caddy are you running?
0.9-beta2

2. What are you trying to do?
Run on Windows.

3. What is your entire Caddyfile?
```text
# PHPInfo
localhost:80 {
	log logs/access.log
	errors logs/error.log
	gzip
	startup php-cgi -b 127.0.0.1:9000 &
	fastcgi / 127.0.0.1:9000 php {
		index index.php
	}
}
```

4. How did you run Caddy?
Renamed `caddy_windows_386.exe` to `caddy.exe`, ran `caddy` in elevated terminal (admin in Cmder).

5. What did you expect to see?
PHPInfo() on localhost:80.

6. What did you see instead?

```
2016/06/23 22:28:56 [ERROR] Unable to migrate certificate storage: rename C:\Users\Ole\.caddy\letsencrypt C:\Users\Ole\.caddy\acme\acme-v01.api.letsencrypt.org: Access is denied.
```

This in spite of `C:\Users\Ole\.caddy` not being "locked" (disabled Read-Only recursively).

7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?
Presumably, download `caddy_windows_386.zip` from release "0.9 beta 2", extract to `C:\Caddy-0.9b2`, rename the .exe and run `caddy`.

## Resolution
Deleted the folder `C:\Users\Ole\.caddy`, and Caddy runs as expected.

**Comment by mholt** [comment]
Thanks for the report! And for trying the beta. I'm not very well versed on Windows so I may need some help here. Administrator didn't have permission to move a folder? How is that possible? :confused:

Deleting the .caddy folder will indeed work, but that will cause certificates to be re-issued and a new account to be registered, so I don't typically recommend it...

**Comment by OleVik** [comment]
It's rather odd, as my account is already administrator and virtually every other CLI I've used can read and write without hindrance. A best guess would be that the creation of `.caddy` results in some permissions set, where `.caddy` (and children) inevitably belong to the user and not the system, hence I can delete but Caddy cannot write.

**Comment by mholt** [comment]
So in Windows, does Administrator not have permissions to write a folder that belongs to a user?

**Comment by OleVik** [comment]
It does generally, because permissions are usually set as belonging to the system and thus any administrator has full control. I'm no expert on the Windows permissions, but a typical issue with them is that either the system or the user has exclusive write-permissions, where the system should be owner and the user have full control. Administrators generally have this full control, except in cases where they are improperly set and the user (Administrator or not) does not get them.

This is often the cause of locked filers/folders, where permissions must be manually changed for the user to delete them - be it through a third-party CLI or manually. I restored `.caddy` to check what the case was, and System, Me, and Administrators all have full control, whereas Application Packages do not. From what I know this is pretty common, so it's hard to tell why Caddy could not move them.

Perhaps there's a Windows guru about who can check out the mechanism by which `.caddy` is created and how it is set to be moved.

**Comment by mholt** [comment]
Well, for now, I'll add instructions to the error message to check out this post to do it manually, it's pretty straightforward and this operation is temporary while people upgrade.

For this guide, assume `~` is the home directory (user folder) on your system. Specifically, it is the path in the `$HOME` / `%HOMEPATH%` / `%USERPROFILE%` variable set in the environment in which Caddy was run.

1. Rename (move) `~/.caddy/letsencrypt` to `~/.caddy/acme`
2. Create a folder inside `~/.caddy/acme` called ~/.caddy/acme/acme-v01.api.letsencrypt.org`
3. Move the `sites` and `users` folders from `~/.caddy/acme` into the new subfolder you created.

We've also assumed that all certificates stored on your system were obtained with the production Let's Encrypt endpoint at `acme-v01.api.letsencrypt.org`. If you obtained any using their staging endpoint, you'll need to follow similar steps for `acme-staging.api.letsencrypt.org`, and mirror the structure in that subfolder only for the accounts and certificates used with that endpoint.

If you do not wish to use `~/.caddy` for storage anymore, you can change its location by changing the `HOME` env variable or you can set `CADDYPATH` and Caddy will use that folder for storage (it will not create a `.caddy` folder inside $CADDYPATH).

**Comment by mholt** [comment]
I was able to reproduce this in Windows and I believe this might be a bug in Go: https://github.com/golang/go/issues/14527

**Comment by mholt** [comment]
@OleVik I also found this issue on Linux! And I think I fixed the problem. filepath.Walk is recursive and pre-loads the list of files to traverse. Renaming one of the folders affected all the others in that part of the tree so by the time we got to those files, their path didn't exist anymore.

Anyway, after putting `filepath.Walk()` into a loop where I restart the walk after every rename, I see the error go away. It will go out with Caddy 0.9. It's slightly slower but I don't expect more than a few files in that folder for most people (and it's a one-time thing).
